### ISSUE TITLE
How to include Google Analytics code only when deploying?

### ISSUE THREAD

**Issue by vitorbaptista** [lektor/lektor#503]
I've set up Google Analytics using a data bag to hold the tracking ID, and a snippet to include in my templates. My problem is: I don't want Google Analytics enabled on development, only when the site is deployed. I looked for a way to detect if Lektor is building or deploying the project, but couldn't find any. Is there a way to do this?

**Comment by goanpeca** [comment]
@vitorbaptista you could perhaps use the extra flag CLI option, then you could use a custom plugin to either pop the GA code or not based on that flag so that it is only taken into account when you passed the flag on deploy?

```
lektor deploy --extra-flag is_deploy
```

or

```
lektor deploy -f deploying
```


See the help

```
lektor deploy --help
...
  -f, --extra-flag TEXT   Defines an arbitrary flag.  These can be used by
                          plugins to customize the build and deploy process.
                          More information can be found in the documentation
                          of affected plugins.
```


---

Maybe its not simplest way but this is definitely an option. Another way of doing this in the future, would be to allow multiple configs and do something on those configs... like enabling disabling plugins or theit options (https://github.com/kmonsoor/lektor-google-analytics)

**Comment by goanpeca** [comment]
@nixjdm maybe we should include by default some flags in the jinja2 environment context if the `deploy` command is executed as opposed to the `build` or the `server` command?

So that users could run some template logic like

```jinja2
{% if DEPLOY %}
...
{% endif %}
```

**Comment by nixjdm** [comment]
I think I like adding DEPLOY or similar into Lektor core. It could be top level like that. Alternatively, we could put it inside something else, like the `config` object that is as yet [undocumented](https://github.com/lektor/lektor-website/issues/87). Does it sound useful to have a more generally named `COMMAND` var whose value could be `deploy`, `server`, etc?

**Comment by goanpeca** [comment]
I am fine with either way I think, @lektor/developers thoughts?

**Comment by vitorbaptista** [comment]
Thanks for the super quick and useful response, guys. I'll try the `--extra-flag`, but in general I think detecting when you're deploying is very useful. Actually, as this is a static builder, the meaning of "deploy" is a bit muddy. It could potentially be "deploy" or "build", as there can be people that simply build the site and deploy outside of `lektor`.

In other words, what I'd really like isn't a deployment flag, but a "not-in-development" flag, as in I'm not running via `lektor server`.

**Comment by goanpeca** [comment]
>In other words, what I'd really like isn't a deployment flag, but a "not-in-development" flag, as in I'm not running via lektor server.

Got it

**Comment by xlotlu** [comment]
I would also like a "development" flag to load assets from the webpack dev server instead.

**Comment by CAM-Gerlach** [comment]
Hey, so I came across this thanks to @xlotlu when I was trying to figure out a way to tell from within a template whether the site is being built interactively via ``lektor server`` or not. The specific problem I'm facing is Lektor injects inline Javascript to enable the link to the admin interface, so unless I enable ``'unsafe-inline'`` for ``script-src`` in CSP (which defeats most of its protection), my theme users won't see the link the admin GUI in their browser and will have to remember and enter the URL manually every time.

If the ``--extra-flags`` approach was as simple as just checking the flag from the template it would be doable, if an extra step for users, but if I understand correctly, it seems I'd have to write a whole custom plugin, get users to install it, *and* configure their build with the proper flag, for the sake of a relatively modest security boost—it just isn't worthwhile.

What I'd like to see—and if its simple, I might be able to help—is just a way for Lektor to somehow expose to the template the command that was run, so I can do something like (the API could be whatever you think best for accessing it) ``{% if env.command == "server" %}{% set script_src  = script_src ~ " unsafe-inline" %}{% endif %}``. This would obviate most of the need for the user setting and checking a bespoke ``extra_flag``, although that could be implemented as well down the road (or as part of this, if its simple enough) for use cases where checking the Lektor command isn't enough for some reason.

Let me know, thanks!

**Comment by nixjdm** [comment]
If we add a top-level global template var `command`, that offers a lot of options like

- `if "server" in command`
- `if "deploy" not in command`
- `if "webpack" in command`

As an implementation detail, we shouldn't literally use the command from the terminal / command prompt, but instead should normalize things a little bit to populate the var with an equivalent string, and not literally the command. That will be more robust. For instance, that might allow the var to be usefully populated if someone deploys through the admin (yielding `deploy`), or use shorthand commands like `lektor s` (yielding `server`).

PRs most welcome :)

--

Edit: Bad example maybe. `deploy` might be problematic since that command doesn't render templates. So think `build` instead.

**Comment by CAM-Gerlach** [comment]
Good point; we want to abstract the command name at least a little bit.

On a higher level, since I've never touched the Lektor codebase before (and am not that pro of a developer in general; I'm an atmospheric scientist after all) I assume this be a pretty straightforward task? Could you give a high-level outline of how you'd want it implemented? Thanks!
